protobuf/lib/src/protobuf/proto3_json.dart

```diff
@@ -89,11 +100,32 @@ Object? _writeToProto3Json(_FieldSet fs, TypeRegistry typeRegistry) {
   var result = <String, dynamic>{};
   for (var fieldInfo in fs._infosSortedByTag) {
     var value = fs._values[fieldInfo.index!];
-    if (value == null || (value is List && value.isEmpty)) {
+    var overrideForEmitsDefaults = false;
+    dynamic overrideForEmitsDefaultsValue;
+    if (context.emitDefaults) {
+      if (fieldInfo.isRepeated && isNullOrEmptyList(value)) {
+        overrideForEmitsDefaults = true;
+        overrideForEmitsDefaultsValue = [];
+      } else if (fieldInfo.isMapField && isNullOrEmptyMap(value)) {
+        overrideForEmitsDefaults = true;
+        overrideForEmitsDefaultsValue = {};
+      } else if (_isBytes(fieldInfo.type) && isNullOrEmptyList(value)) {
+        overrideForEmitsDefaults = true;
+        overrideForEmitsDefaultsValue = null;
+      } else if (_isGroupOrMessage(fieldInfo.type) && value == null) {
```

Comment:
For message fields null check should not give us the right answer here. See https://github.com/google/protobuf.dart/issues/309 for details.

Interestingly, I think you have a test case for this (`testToProto3JsonEmitDefaultsWithChild`) and it somehow passes.. I wonder how that works..

Reply:
> Interestingly, I think you have a test case for this (testToProto3JsonEmitDefaultsWithChild) and it somehow passes.. I wonder how that works..

Reading that test again, the child message is not the same as the default value for the field, so it *does not* test that the `emitDefaults` flag works as expected when a field with a message type has the default value.

Could you update the test to check that with `emitDefault: true` we don't serialize a message field if the value is the same as the default?

Note that currently default value for messages are not `null` in protobuf.dart, see #309.

Reply:
Great feedback.  I'm going to create a new set of tests for this that I think will be more comprehensive and help ensure all uses cases are properly.  I'll get started on that now.

Reply:
I have finished adding the additional tests.  The tests confirm that the serialization is behaving as one would expect.  Please have a look and see if you're aligned as well.